cmd/admin-router.go

```diff
@@ -27,50 +35,54 @@ func registerAdminRouter(mux *router.Router) {
+	adminV1Router.Methods(http.MethodDelete).Path("/locks").HandlerFunc(adminAPI.ClearLocksHandler)
 
 	/// Heal operations
 
-	// List Objects needing heal.
-	adminRouter.Methods("GET").Queries("heal", "").Headers(minioAdminOpHeader, "list-objects").HandlerFunc(adminAPI.ListObjectsHealHandler)
-	// List Uploads needing heal.
-	adminRouter.Methods("GET").Queries("heal", "").Headers(minioAdminOpHeader, "list-uploads").HandlerFunc(adminAPI.ListUploadsHealHandler)
-	// List Buckets needing heal.
-	adminRouter.Methods("GET").Queries("heal", "").Headers(minioAdminOpHeader, "list-buckets").HandlerFunc(adminAPI.ListBucketsHealHandler)
-
-	// Heal Buckets.
-	adminRouter.Methods("POST").Queries("heal", "").Headers(minioAdminOpHeader, "bucket").HandlerFunc(adminAPI.HealBucketHandler)
-	// Heal Objects.
-	adminRouter.Methods("POST").Queries("heal", "").Headers(minioAdminOpHeader, "object").HandlerFunc(adminAPI.HealObjectHandler)
-	// Heal Format.
-	adminRouter.Methods("POST").Queries("heal", "").Headers(minioAdminOpHeader, "format").HandlerFunc(adminAPI.HealFormatHandler)
-	// Heal Uploads.
-	adminRouter.Methods("POST").Queries("heal", "").Headers(minioAdminOpHeader, "upload").HandlerFunc(adminAPI.HealUploadHandler)
+	// Fetch and return dynamic heal status.
+	adminV1Router.Methods(http.MethodGet).Path("/heal/").HandlerFunc(adminAPI.HealStatusHandler)
+	adminV1Router.Methods(http.MethodGet).Path("/heal/{bucket}").HandlerFunc(adminAPI.HealStatusHandler)
+	adminV1Router.Methods(http.MethodGet).Path("/heal/{bucket}/{prefix:.*}").HandlerFunc(adminAPI.HealStatusHandler)
+
+	// Start heal operation.
+	adminV1Router.Methods(http.MethodPost).Path("/heal/").HandlerFunc(adminAPI.HealStartHandler)
+	adminV1Router.Methods(http.MethodPost).Path("/heal/{bucket}").HandlerFunc(adminAPI.HealStartHandler)
+	adminV1Router.Methods(http.MethodPost).Path("/heal/{bucket}/{prefix:.*}").HandlerFunc(adminAPI.HealStartHandler)
+
+	// Stop a healing operation.
+	adminV1Router.Methods(http.MethodDelete).Path("/heal/").HandlerFunc(adminAPI.HealStopHandler)
+	adminV1Router.Methods(http.MethodDelete).Path("/heal/{bucket}").HandlerFunc(adminAPI.HealStopHandler)
+	adminV1Router.Methods(http.MethodDelete).Path("/heal/{bucket}/{prefix:.*}").HandlerFunc(adminAPI.HealStopHandler)
 
 	/// Config operations
 
+	// Update credentials
+	adminV1Router.Methods(http.MethodPost).Path("/config/credential").HandlerFunc(adminAPI.UpdateCredentialsHandler)
 	// Get config
-	adminRouter.Methods("GET").Queries("config", "").Headers(minioAdminOpHeader, "get").HandlerFunc(adminAPI.GetConfigHandler)
-	// Set Config
-	adminRouter.Methods("PUT").Queries("config", "").Headers(minioAdminOpHeader, "set").HandlerFunc(adminAPI.SetConfigHandler)
+	adminV1Router.Methods(http.MethodGet).Path("/config").HandlerFunc(adminAPI.GetConfigHandler)
+	// Set config
+	adminV1Router.Methods(http.MethodPut).Path("/config").HandlerFunc(adminAPI.SetConfigHandler)
```

Comment:
the spec https://gist.github.com/donatello/e520308bf83adddc70366ea01ef505e5 says POST for this.

POST will be good as the calls are either GET or POST or DELETE. There is no PUT, so we'll never get confused on if a call is PUT.

Reply:
@krishnasrinivas PUT seems to me to be appropriate here because it completely replaces the configuration and is safe to repeat (it is idempotent) (if the config update succeeds at least once, it can be repeated with no ill-effects) - this is not in general true of POST. I think the issue of confusing with POST is a somewhat lesser one because the spec will be corrected and most users will use the client tools or the madmin package.